### ISSUE TITLE
[Bug]:TypeError: 'generator' object does not support the context manager protocol

### ISSUE THREAD

**Issue by mike2505** [openlit/openlit#391]
### 🐛 What's Going Wrong?
I'm encountering an error when attempting to use agent_executor.invoke() in my application with OpenLIT. The error occurs during the invocation of the agent and appears to be related to a generator object not supporting the context manager protocol.

### 🕵️ Steps to Reproduce
Add OpenLIT to the application via:
```
openlit.init(
    otlp_endpoint="http://127.0.0.1:4318",
)
```
Set up the AgentExecutor:
```
agent_executor = AgentExecutor(agent=agent, tools=functions, verbose=True)
```
Attempt to invoke the agent with the following code:
```
agent_output = agent_executor.invoke(
    {
        "input": message,
        "chat_history": chat_history,
    }
)
```
The error is raised on this line during execution.

### 🎯 What Did You Expect?
I expected the agent_executor.invoke() function to run without errors, properly invoking the agent and returning the expected output.

### 📸 Any Screenshots?
Unfortunately, I don't have any relevant screenshots at the moment, but I have provided the traceback below.

### 📝 Additional Notes
The full traceback of the error is as follows:

```File "/home/mikef0x/slots.gd/app/utils/llm.py", line 78, in llm_response_openai
    agent_output = agent_executor.invoke(
                   ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain/chains/base.py", line 164, in invoke
    raise e
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain/chains/base.py", line 154, in invoke
    self._call(inputs, run_manager=run_manager)
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain/agents/agent.py", line 1608, in _call
    next_step_output = self._take_next_step(
                       ^^^^^^^^^^^^^^^^^^^^^
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain/agents/agent.py", line 1314, in _take_next_step
    [
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain/agents/agent.py", line 1342, in _iter_next_step
    output = self.agent.plan(
             ^^^^^^^^^^^^^^^^
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain/agents/agent.py", line 576, in plan
    for chunk in self.runnable.stream(inputs, config={"callbacks": callbacks}):
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/runnables/base.py", line 3262, in stream
    yield from self.transform(iter([input]), config, **kwargs)
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/runnables/base.py", line 3249, in transform
    yield from self._transform_stream_with_config(
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/runnables/base.py", line 2054, in _transform_stream_with_config
    chunk: Output = context.run(next, iterator)  # type: ignore
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/runnables/base.py", line 3211, in _transform
    for output in final_pipeline:
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/runnables/base.py", line 1272, in transform
    for ichunk in input:
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/runnables/base.py", line 5301, in transform
    yield from self.bound.transform(
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/runnables/base.py", line 1290, in transform
    yield from self.stream(final, config, **kwargs)
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/language_models/chat_models.py", line 425, in stream
    raise e
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_core/language_models/chat_models.py", line 405, in stream
    for chunk in self._stream(messages, stop=stop, **kwargs):
  File "/home/mikef0x/slots.gd/env/lib/python3.12/site-packages/langchain_openai/chat_models/base.py", line 559, in _stream
    with response:
TypeError: 'generator' object does not support the context manager protocol```

**Comment by masnec** [comment]
Run into the same issue, it's caused by the context manager change in langchain - https://github.com/langchain-ai/langchain/pull/18013 . May need a fix in https://github.com/openlit/openlit/blob/main/sdk/python/src/openlit/instrumentation/openai/async_openai.py

**Comment by amanagarwal042** [comment]
Ack
I am looking for a fix on this as directly openai works but calling an agent in this way fails. Ill keep you posted on this

Any ideas @masnec why you think async openai is being used here?

**Comment by masnec** [comment]
@patcher9 the issue posted here doesn't use async openai. But I think we need a fix for both sync and async flows since https://github.com/langchain-ai/langchain/pull/18013 updated both code paths.

**Comment by amanagarwal042** [comment]
Alright thanks, This might mean we would need a different way to capture the streaming response. Thanks for pointing it out

**Comment by zacchenit** [comment]
Could you please provide any updates on this issue?

**Comment by amanagarwal042** [comment]
> Could you please provide any updates on this issue?

I think I have a way now to call the streaming object, Just need to make sure its working with all frameworks

**Comment by amanagarwal042** [comment]
Changing the method seems to break it for other frameworks.

[Though I just noticed AgentExecutor is lagcy now and LangChain recommends using LangGraph to create agents](https://python.langchain.com/docs/how_to/agent_executor/)

So to test if OpenLIT SDK works with the recommended method, I tried out the below code


```python
# Import relevant functionality
from langchain_openai import ChatOpenAI
from langchain_community.tools.tavily_search import TavilySearchResults
from langchain_core.messages import HumanMessage
from langgraph.checkpoint.memory import MemorySaver
from langgraph.prebuilt import create_react_agent
import openlit

openlit.init()

# Create the agent
memory = MemorySaver()
model = ChatOpenAI(model_name="gpt-4")
search = TavilySearchResults(max_results=2)
tools = [search]
agent_executor = create_react_agent(model, tools, checkpointer=memory)

# Use the agent
config = {"configurable": {"thread_id": "abc123"}}
for chunk in agent_executor.stream(
    {"messages": [HumanMessage(content="hi im bob! and i live in sf")]}, config
):
    print(chunk)
    print("----")

for chunk in agent_executor.stream(
    {"messages": [HumanMessage(content="whats the weather where I live?")]}, config
):
    print(chunk)
    print("----")
```

and it seems to work as expected. See the below generated trace

<details>
<summary>OTel Trace and Metrics printed on cosnole</summary>

```json
{
    "name": "openai.chat.completions",
    "context": {
        "trace_id": "0x60cae6be02cfec712320a4402ed91e5b",
        "span_id": "0x5943e331d143ba0f",
        "trace_state": "[]"
    },
    "kind": "SpanKind.CLIENT",
    "parent_id": "0x0b7cd4e4c9491df7",
    "start_time": "2024-10-17T14:00:05.006830Z",
    "end_time": "2024-10-17T14:00:05.943427Z",
    "status": {
        "status_code": "OK"
    },
    "attributes": {
        "telemetry.sdk.name": "openlit",
        "gen_ai.system": "openai",
        "gen_ai.operation.name": "chat",
        "gen_ai.endpoint": "openai.chat.completions",
        "gen_ai.response.id": "chatcmpl-AJLKnhFt4qvnp3FxmkhHAlPssDmgK",
        "gen_ai.environment": "default",
        "gen_ai.application_name": "default",
        "gen_ai.request.model": "gpt-4",
        "gen_ai.request.top_p": 1.0,
        "gen_ai.request.max_tokens": -1,
        "gen_ai.request.user": "",
        "gen_ai.request.temperature": 0.7,
        "gen_ai.request.presence_penalty": 0.0,
        "gen_ai.request.frequency_penalty": 0.0,
        "gen_ai.request.seed": "",
        "gen_ai.request.is_stream": false,
        "gen_ai.usage.input_tokens": 90,
        "gen_ai.usage.output_tokens": 11,
        "gen_ai.usage.total_tokens": 101,
        "gen_ai.usage.cost": 0.0033599999999999997
    },
    "events": [
        {
            "name": "gen_ai.content.prompt",
            "timestamp": "2024-10-17T14:00:05.942941Z",
            "attributes": {
                "gen_ai.prompt": "user: hi im bob! and i live in sf"
            }
        },
        {
            "name": "gen_ai.content.completion",
            "timestamp": "2024-10-17T14:00:05.943010Z",
            "attributes": {
                "gen_ai.completion": "Function called with tools"
            }
        }
    ],
    "links": [],
    "resource": {
        "attributes": {
            "telemetry.sdk.language": "python",
            "telemetry.sdk.name": "openlit",
            "telemetry.sdk.version": "1.27.0",
            "service.name": "default",
            "deployment.environment": "default"
        },
        "schema_url": ""
    }
}
{
    "name": "langchain.chat_models",
    "context": {
        "trace_id": "0x60cae6be02cfec712320a4402ed91e5b",
        "span_id": "0x0b7cd4e4c9491df7",
        "trace_state": "[]"
    },
    "kind": "SpanKind.CLIENT",
    "parent_id": null,
    "start_time": "2024-10-17T14:00:05.006478Z",
    "end_time": "2024-10-17T14:00:05.950981Z",
    "status": {
        "status_code": "OK"
    },
    "attributes": {
        "telemetry.sdk.name": "openlit",
        "gen_ai.system": "langchain",
        "gen_ai.endpoint": "langchain.chat_models",
        "gen_ai.environment": "default",
        "gen_ai.operation.name": "chat",
        "gen_ai.application_name": "default",
        "gen_ai.request.model": "gpt-4",
        "gen_ai.request.temperature": "0.7",
        "gen_ai.request.top_k": "1",
        "gen_ai.request.top_p": "None",
        "gen_ai.request.is_stream": false,
        "gen_ai.usage.input_tokens": 0,
        "gen_ai.usage.output_tokens": 0,
        "gen_ai.usage.total_tokens": 0,
        "gen_ai.usage.cost": 0.0
    },
    "events": [
        {
            "name": "gen_ai.content.prompt",
            "timestamp": "2024-10-17T14:00:05.950879Z",
            "attributes": {
                "gen_ai.prompt": ""
            }
        },
        {
            "name": "gen_ai.content.completion",
            "timestamp": "2024-10-17T14:00:05.950889Z",
            "attributes": {
                "gen_ai.completion": "Hello Bob! How can I assist you today?"
            }
        }
    ],
    "links": [],
    "resource": {
        "attributes": {
            "telemetry.sdk.language": "python",
            "telemetry.sdk.name": "openlit",
            "telemetry.sdk.version": "1.27.0",
            "service.name": "default",
            "deployment.environment": "default"
        },
        "schema_url": ""
    }
}
{'agent': {'messages': [AIMessage(content='Hello Bob! How can I assist you today?', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 11, 'prompt_tokens': 90, 'total_tokens': 101, 'completion_tokens_details': {'audio_tokens': None, 'reasoning_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': None, 'cached_tokens': 0}}, 'model_name': 'gpt-4-0613', 'system_fingerprint': None, 'finish_reason': 'stop', 'logprobs': None}, id='run-a87ca98f-9db2-4e0f-9f72-8cae601d7da9-0', usage_metadata={'input_tokens': 90, 'output_tokens': 11, 'total_tokens': 101, 'input_token_details': {'cache_read': 0}, 'output_token_details': {'reasoning': 0}})]}}
----
{
    "name": "openai.chat.completions",
    "context": {
        "trace_id": "0x4c34ed3188c9dfed03ff581db769a9a3",
        "span_id": "0xe7524fed4a81422d",
        "trace_state": "[]"
    },
    "kind": "SpanKind.CLIENT",
    "parent_id": "0x108b689ed75ca938",
    "start_time": "2024-10-17T14:00:05.956832Z",
    "end_time": "2024-10-17T14:00:07.415606Z",
    "status": {
        "status_code": "OK"
    },
    "attributes": {
        "telemetry.sdk.name": "openlit",
        "gen_ai.system": "openai",
        "gen_ai.operation.name": "chat",
        "gen_ai.endpoint": "openai.chat.completions",
        "gen_ai.response.id": "chatcmpl-AJLKoZyVVd8kKuIZBQ1iU9tQYDy1t",
        "gen_ai.environment": "default",
        "gen_ai.application_name": "default",
        "gen_ai.request.model": "gpt-4",
        "gen_ai.request.top_p": 1.0,
        "gen_ai.request.max_tokens": -1,
        "gen_ai.request.user": "",
        "gen_ai.request.temperature": 0.7,
        "gen_ai.request.presence_penalty": 0.0,
        "gen_ai.request.frequency_penalty": 0.0,
        "gen_ai.request.seed": "",
        "gen_ai.request.is_stream": false,
        "gen_ai.usage.input_tokens": 116,
        "gen_ai.usage.output_tokens": 23,
        "gen_ai.usage.total_tokens": 139,
        "gen_ai.usage.cost": 0.00486
    },
    "events": [
        {
            "name": "gen_ai.content.prompt",
            "timestamp": "2024-10-17T14:00:07.415497Z",
            "attributes": {
                "gen_ai.prompt": "user: hi im bob! and i live in sf\nassistant: Hello Bob! How can I assist you today?\nuser: whats the weather where I live?"
            }
        },
        {
            "name": "gen_ai.content.completion",
            "timestamp": "2024-10-17T14:00:07.415515Z",
            "attributes": {
                "gen_ai.completion": "Function called with tools"
            }
        }
    ],
    "links": [],
    "resource": {
        "attributes": {
            "telemetry.sdk.language": "python",
            "telemetry.sdk.name": "openlit",
            "telemetry.sdk.version": "1.27.0",
            "service.name": "default",
            "deployment.environment": "default"
        },
        "schema_url": ""
    }
}
{
    "name": "langchain.chat_models",
    "context": {
        "trace_id": "0x4c34ed3188c9dfed03ff581db769a9a3",
        "span_id": "0x108b689ed75ca938",
        "trace_state": "[]"
    },
    "kind": "SpanKind.CLIENT",
    "parent_id": null,
    "start_time": "2024-10-17T14:00:05.956407Z",
    "end_time": "2024-10-17T14:00:07.417071Z",
    "status": {
        "status_code": "OK"
    },
    "attributes": {
        "telemetry.sdk.name": "openlit",
        "gen_ai.system": "langchain",
        "gen_ai.endpoint": "langchain.chat_models",
        "gen_ai.environment": "default",
        "gen_ai.operation.name": "chat",
        "gen_ai.application_name": "default",
        "gen_ai.request.model": "gpt-4",
        "gen_ai.request.temperature": "0.7",
        "gen_ai.request.top_k": "1",
        "gen_ai.request.top_p": "None",
        "gen_ai.request.is_stream": false,
        "gen_ai.usage.input_tokens": 0,
        "gen_ai.usage.output_tokens": 0,
        "gen_ai.usage.total_tokens": 0,
        "gen_ai.usage.cost": 0.0
    },
    "events": [
        {
            "name": "gen_ai.content.prompt",
            "timestamp": "2024-10-17T14:00:07.417027Z",
            "attributes": {
                "gen_ai.prompt": ""
            }
        },
        {
            "name": "gen_ai.content.completion",
            "timestamp": "2024-10-17T14:00:07.417036Z",
            "attributes": {
                "gen_ai.completion": ""
            }
        }
    ],
    "links": [],
    "resource": {
        "attributes": {
            "telemetry.sdk.language": "python",
            "telemetry.sdk.name": "openlit",
            "telemetry.sdk.version": "1.27.0",
            "service.name": "default",
            "deployment.environment": "default"
        },
        "schema_url": ""
    }
}
{'agent': {'messages': [AIMessage(content='', additional_kwargs={'tool_calls': [{'id': 'call_yZ270GJ8qZSNHiunGki7pNMN', 'function': {'arguments': '{\n  "query": "current weather in San Francisco"\n}', 'name': 'tavily_search_results_json'}, 'type': 'function'}], 'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 23, 'prompt_tokens': 116, 'total_tokens': 139, 'completion_tokens_details': {'audio_tokens': None, 'reasoning_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': None, 'cached_tokens': 0}}, 'model_name': 'gpt-4-0613', 'system_fingerprint': None, 'finish_reason': 'tool_calls', 'logprobs': None}, id='run-8edfc9d4-2493-4df2-8ee8-d2b7dac489ff-0', tool_calls=[{'name': 'tavily_search_results_json', 'args': {'query': 'current weather in San Francisco'}, 'id': 'call_yZ270GJ8qZSNHiunGki7pNMN', 'type': 'tool_call'}], usage_metadata={'input_tokens': 116, 'output_tokens': 23, 'total_tokens': 139, 'input_token_details': {'cache_read': 0}, 'output_token_details': {'reasoning': 0}})]}}
----
{'tools': {'messages': [ToolMessage(content='[{"url": "https://www.weatherapi.com/", "content": "{\'location\': {\'name\': \'San Francisco\', \'region\': \'California\', \'country\': \'United States of America\', \'lat\': 37.775, \'lon\': -122.4183, \'tz_id\': \'America/Los_Angeles\', \'localtime_epoch\': 1729173528, \'localtime\': \'2024-10-17 06:58\'}, \'current\': {\'last_updated_epoch\': 1729172700, \'last_updated\': \'2024-10-17 06:45\', \'temp_c\': 13.3, \'temp_f\': 55.9, \'is_day\': 0, \'condition\': {\'text\': \'Partly cloudy\', \'icon\': \'//cdn.weatherapi.com/weather/64x64/night/116.png\', \'code\': 1003}, \'wind_mph\': 8.5, \'wind_kph\': 13.7, \'wind_degree\': 280, \'wind_dir\': \'W\', \'pressure_mb\': 1013.0, \'pressure_in\': 29.9, \'precip_mm\': 0.0, \'precip_in\': 0.0, \'humidity\': 87, \'cloud\': 50, \'feelslike_c\': 12.1, \'feelslike_f\': 53.8, \'windchill_c\': 10.2, \'windchill_f\': 50.4, \'heatindex_c\': 11.6, \'heatindex_f\': 52.9, \'dewpoint_c\': 11.2, \'dewpoint_f\': 52.2, \'vis_km\': 16.0, \'vis_miles\': 9.0, \'uv\': 0.0, \'gust_mph\': 12.3, \'gust_kph\': 19.8}}"}, {"url": "https://www.weathertab.com/en/c/e/10/united-states/california/san-francisco/", "content": "Explore comprehensive October 2024 weather forecasts for San Francisco, including daily high and low temperatures, precipitation risks, and monthly temperature trends. Featuring detailed day-by-day forecasts, dynamic graphs of daily rain probabilities, and temperature trends to help you plan ahead. ... 17 66°F 54°F 19°C 12°C 27% 18 66°F 53"}]', name='tavily_search_results_json', id='38c01532-214d-40fc-979f-1c8fd49d42f3', tool_call_id='call_yZ270GJ8qZSNHiunGki7pNMN', artifact={'query': 'current weather in San Francisco', 'follow_up_questions': None, 'answer': None, 'images': [], 'results': [{'title': 'Weather in San Francisco', 'url': 'https://www.weatherapi.com/', 'content': "{'location': {'name': 'San Francisco', 'region': 'California', 'country': 'United States of America', 'lat': 37.775, 'lon': -122.4183, 'tz_id': 'America/Los_Angeles', 'localtime_epoch': 1729173528, 'localtime': '2024-10-17 06:58'}, 'current': {'last_updated_epoch': 1729172700, 'last_updated': '2024-10-17 06:45', 'temp_c': 13.3, 'temp_f': 55.9, 'is_day': 0, 'condition': {'text': 'Partly cloudy', 'icon': '//cdn.weatherapi.com/weather/64x64/night/116.png', 'code': 1003}, 'wind_mph': 8.5, 'wind_kph': 13.7, 'wind_degree': 280, 'wind_dir': 'W', 'pressure_mb': 1013.0, 'pressure_in': 29.9, 'precip_mm': 0.0, 'precip_in': 0.0, 'humidity': 87, 'cloud': 50, 'feelslike_c': 12.1, 'feelslike_f': 53.8, 'windchill_c': 10.2, 'windchill_f': 50.4, 'heatindex_c': 11.6, 'heatindex_f': 52.9, 'dewpoint_c': 11.2, 'dewpoint_f': 52.2, 'vis_km': 16.0, 'vis_miles': 9.0, 'uv': 0.0, 'gust_mph': 12.3, 'gust_kph': 19.8}}", 'score': 0.9914166, 'raw_content': None}, {'title': 'San Francisco, CA Weather Forecast October 2024: Daily Highs/Lows ...', 'url': 'https://www.weathertab.com/en/c/e/10/united-states/california/san-francisco/', 'content': 'Explore comprehensive October 2024 weather forecasts for San Francisco, including daily high and low temperatures, precipitation risks, and monthly temperature trends. Featuring detailed day-by-day forecasts, dynamic graphs of daily rain probabilities, and temperature trends to help you plan ahead. ... 17 66°F 54°F 19°C 12°C 27% 18 66°F 53', 'score': 0.9893885, 'raw_content': None}], 'response_time': 3.72})]}}
----
{
    "name": "openai.chat.completions",
    "context": {
        "trace_id": "0x07e8067bb598e8b4e52bcafa8b9ac0f7",
        "span_id": "0x9dd63eb52579d884",
        "trace_state": "[]"
    },
    "kind": "SpanKind.CLIENT",
    "parent_id": "0x9481b7a670f0ae68",
    "start_time": "2024-10-17T14:00:11.426790Z",
    "end_time": "2024-10-17T14:00:14.220863Z",
    "status": {
        "status_code": "OK"
    },
    "attributes": {
        "telemetry.sdk.name": "openlit",
        "gen_ai.system": "openai",
        "gen_ai.operation.name": "chat",
        "gen_ai.endpoint": "openai.chat.completions",
        "gen_ai.response.id": "chatcmpl-AJLKtG3vc1nvClXJzwOUO7Zsr7Y5p",
        "gen_ai.environment": "default",
        "gen_ai.application_name": "default",
        "gen_ai.request.model": "gpt-4",
        "gen_ai.request.top_p": 1.0,
        "gen_ai.request.max_tokens": -1,
        "gen_ai.request.user": "",
        "gen_ai.request.temperature": 0.7,
        "gen_ai.request.presence_penalty": 0.0,
        "gen_ai.request.frequency_penalty": 0.0,
        "gen_ai.request.seed": "",
        "gen_ai.request.is_stream": false,
        "gen_ai.usage.input_tokens": 667,
        "gen_ai.usage.output_tokens": 49,
        "gen_ai.usage.total_tokens": 716,
        "gen_ai.usage.cost": 0.022949999999999998
    },
    "events": [
        {
            "name": "gen_ai.content.prompt",
            "timestamp": "2024-10-17T14:00:14.220769Z",
            "attributes": {
                "gen_ai.prompt": "user: hi im bob! and i live in sf\nassistant: Hello Bob! How can I assist you today?\nuser: whats the weather where I live?\nassistant: None\ntool: [{\"url\": \"https://www.weatherapi.com/\", \"content\": \"{'location': {'name': 'San Francisco', 'region': 'California', 'country': 'United States of America', 'lat': 37.775, 'lon': -122.4183, 'tz_id': 'America/Los_Angeles', 'localtime_epoch': 1729173528, 'localtime': '2024-10-17 06:58'}, 'current': {'last_updated_epoch': 1729172700, 'last_updated': '2024-10-17 06:45', 'temp_c': 13.3, 'temp_f': 55.9, 'is_day': 0, 'condition': {'text': 'Partly cloudy', 'icon': '//cdn.weatherapi.com/weather/64x64/night/116.png', 'code': 1003}, 'wind_mph': 8.5, 'wind_kph': 13.7, 'wind_degree': 280, 'wind_dir': 'W', 'pressure_mb': 1013.0, 'pressure_in': 29.9, 'precip_mm': 0.0, 'precip_in': 0.0, 'humidity': 87, 'cloud': 50, 'feelslike_c': 12.1, 'feelslike_f': 53.8, 'windchill_c': 10.2, 'windchill_f': 50.4, 'heatindex_c': 11.6, 'heatindex_f': 52.9, 'dewpoint_c': 11.2, 'dewpoint_f': 52.2, 'vis_km': 16.0, 'vis_miles': 9.0, 'uv': 0.0, 'gust_mph': 12.3, 'gust_kph': 19.8}}\"}, {\"url\": \"https://www.weathertab.com/en/c/e/10/united-states/california/san-francisco/\", \"content\": \"Explore comprehensive October 2024 weather forecasts for San Francisco, including daily high and low temperatures, precipitation risks, and monthly temperature trends. Featuring detailed day-by-day forecasts, dynamic graphs of daily rain probabilities, and temperature trends to help you plan ahead. ... 17 66\u00b0F 54\u00b0F 19\u00b0C 12\u00b0C 27% 18 66\u00b0F 53\"}]"
            }
        },
        {
            "name": "gen_ai.content.completion",
            "timestamp": "2024-10-17T14:00:14.220786Z",
            "attributes": {
                "gen_ai.completion": "Function called with tools"
            }
        }
    ],
    "links": [],
    "resource": {
        "attributes": {
            "telemetry.sdk.language": "python",
            "telemetry.sdk.name": "openlit",
            "telemetry.sdk.version": "1.27.0",
            "service.name": "default",
            "deployment.environment": "default"
        },
        "schema_url": ""
    }
}
{
    "name": "langchain.chat_models",
    "context": {
        "trace_id": "0x07e8067bb598e8b4e52bcafa8b9ac0f7",
        "span_id": "0x9481b7a670f0ae68",
        "trace_state": "[]"
    },
    "kind": "SpanKind.CLIENT",
    "parent_id": null,
    "start_time": "2024-10-17T14:00:11.426518Z",
    "end_time": "2024-10-17T14:00:14.222233Z",
    "status": {
        "status_code": "OK"
    },
    "attributes": {
        "telemetry.sdk.name": "openlit",
        "gen_ai.system": "langchain",
        "gen_ai.endpoint": "langchain.chat_models",
        "gen_ai.environment": "default",
        "gen_ai.operation.name": "chat",
        "gen_ai.application_name": "default",
        "gen_ai.request.model": "gpt-4",
        "gen_ai.request.temperature": "0.7",
        "gen_ai.request.top_k": "1",
        "gen_ai.request.top_p": "None",
        "gen_ai.request.is_stream": false,
        "gen_ai.usage.input_tokens": 0,
        "gen_ai.usage.output_tokens": 0,
        "gen_ai.usage.total_tokens": 0,
        "gen_ai.usage.cost": 0.0
    },
    "events": [
        {
            "name": "gen_ai.content.prompt",
            "timestamp": "2024-10-17T14:00:14.222173Z",
            "attributes": {
                "gen_ai.prompt": ""
            }
        },
        {
            "name": "gen_ai.content.completion",
            "timestamp": "2024-10-17T14:00:14.222182Z",
            "attributes": {
                "gen_ai.completion": "The current weather in San Francisco is partly cloudy with a temperature of 13.3\u00b0C (55.9\u00b0F). The wind is coming from the west at a speed of 8.5 mph. The humidity level is 87%."
            }
        }
    ],
    "links": [],
    "resource": {
        "attributes": {
            "telemetry.sdk.language": "python",
            "telemetry.sdk.name": "openlit",
            "telemetry.sdk.version": "1.27.0",
            "service.name": "default",
            "deployment.environment": "default"
        },
        "schema_url": ""
    }
}
{'agent': {'messages': [AIMessage(content='The current weather in San Francisco is partly cloudy with a temperature of 13.3°C (55.9°F). The wind is coming from the west at a speed of 8.5 mph. The humidity level is 87%.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 49, 'prompt_tokens': 667, 'total_tokens': 716, 'completion_tokens_details': {'audio_tokens': None, 'reasoning_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': None, 'cached_tokens': 0}}, 'model_name': 'gpt-4-0613', 'system_fingerprint': None, 'finish_reason': 'stop', 'logprobs': None}, id='run-eba6c1b3-72aa-47c3-a705-95a8cced4fee-0', usage_metadata={'input_tokens': 667, 'output_tokens': 49, 'total_tokens': 716, 'input_token_details': {'cache_read': 0}, 'output_token_details': {'reasoning': 0}})]}}
----
{
    "resource_metrics": [
        {
            "resource": {
                "attributes": {
                    "telemetry.sdk.language": "python",
                    "telemetry.sdk.name": "openlit",
                    "telemetry.sdk.version": "1.27.0",
                    "service.name": "default",
                    "deployment.environment": "default"
                },
                "schema_url": ""
            },
            "scope_metrics": [
                {
                    "scope": {
                        "name": "openlit.otel.metrics",
                        "version": "0.1.0",
                        "schema_url": "",
                        "attributes": null
                    },
                    "metrics": [
                        {
                            "name": "gen_ai.total.requests",
                            "description": "Number of requests to GenAI",
                            "unit": "1",
                            "data": {
                                "data_points": [
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "openai",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605943211000,
                                        "time_unix_nano": 1729173614224164000,
                                        "value": 3
                                    },
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "langchain",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605950916000,
                                        "time_unix_nano": 1729173614224164000,
                                        "value": 3
                                    }
                                ],
                                "aggregation_temporality": 2,
                                "is_monotonic": true
                            }
                        },
                        {
                            "name": "gen_ai.usage.total_tokens",
                            "description": "Number of total tokens processed.",
                            "unit": "1",
                            "data": {
                                "data_points": [
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "openai",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605943247000,
                                        "time_unix_nano": 1729173614224164000,
                                        "value": 956
                                    },
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "langchain",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605950930000,
                                        "time_unix_nano": 1729173614224164000,
                                        "value": 0
                                    }
                                ],
                                "aggregation_temporality": 2,
                                "is_monotonic": true
                            }
                        },
                        {
                            "name": "gen_ai.usage.output_tokens",
                            "description": "Number of completion tokens processed.",
                            "unit": "1",
                            "data": {
                                "data_points": [
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "openai",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605943264000,
                                        "time_unix_nano": 1729173614224164000,
                                        "value": 83
                                    },
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "langchain",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605950938000,
                                        "time_unix_nano": 1729173614224164000,
                                        "value": 0
                                    }
                                ],
                                "aggregation_temporality": 2,
                                "is_monotonic": true
                            }
                        },
                        {
                            "name": "gen_ai.usage.input_tokens",
                            "description": "Number of prompt tokens processed.",
                            "unit": "1",
                            "data": {
                                "data_points": [
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "openai",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605943278000,
                                        "time_unix_nano": 1729173614224164000,
                                        "value": 873
                                    },
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "langchain",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605950944000,
                                        "time_unix_nano": 1729173614224164000,
                                        "value": 0
                                    }
                                ],
                                "aggregation_temporality": 2,
                                "is_monotonic": true
                            }
                        },
                        {
                            "name": "gen_ai.usage.cost",
                            "description": "The distribution of GenAI request costs.",
                            "unit": "USD",
                            "data": {
                                "data_points": [
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "openai",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605943365000,
                                        "time_unix_nano": 1729173614224164000,
                                        "count": 3,
                                        "sum": 0.031169999999999996,
                                        "bucket_counts": [
                                            0,
                                            3,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0
                                        ],
                                        "explicit_bounds": [
                                            0.0,
                                            5.0,
                                            10.0,
                                            25.0,
                                            50.0,
                                            75.0,
                                            100.0,
                                            250.0,
                                            500.0,
                                            750.0,
                                            1000.0,
                                            2500.0,
                                            5000.0,
                                            7500.0,
                                            10000.0
                                        ],
                                        "min": 0.0033599999999999997,
                                        "max": 0.022949999999999998
                                    },
                                    {
                                        "attributes": {
                                            "telemetry.sdk.name": "openlit",
                                            "gen_ai.application_name": "default",
                                            "gen_ai.system": "langchain",
                                            "gen_ai.environment": "default",
                                            "gen_ai.operation.name": "chat",
                                            "gen_ai.request.model": "gpt-4"
                                        },
                                        "start_time_unix_nano": 1729173605950953000,
                                        "time_unix_nano": 1729173614224164000,
                                        "count": 3,
                                        "sum": 0.0,
                                        "bucket_counts": [
                                            3,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0,
                                            0
                                        ],
                                        "explicit_bounds": [
                                            0.0,
                                            5.0,
                                            10.0,
                                            25.0,
                                            50.0,
                                            75.0,
                                            100.0,
                                            250.0,
                                            500.0,
                                            750.0,
                                            1000.0,
                                            2500.0,
                                            5000.0,
                                            7500.0,
                                            10000.0
                                        ],
                                        "min": 0.0,
                                        "max": 0.0
                                    }
                                ],
                                "aggregation_temporality": 2
                            }
                        }
                    ],
                    "schema_url": ""
                }
            ],
            "schema_url": ""
        }
    ]
}
```
</details>


Since the recommended way works, a bit hesitant to support something thats legacy

**Comment by MohanadAbugharbia** [comment]
Hello, I believe that I have an issue of a similar/same cause

python 3.11
openlit 1.29.0
langchai-openai 0.2.6
openai 1.54.3

```python
from openai import AsyncOpenAI
from langchain_openai import ChatOpenAI
import openlit

openlit.init(otlp_endpoint="http://127.0.0.1:4318")

chain = ChatOpenAI(
    model="gpt-3.5-turbo",
)

async def get_chat_response(question):
    completion = await chain.astream(
        [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": question},
        ],
    )
    async for response in completion:
        yield response


async def main():
    async for response in get_chat_response("What is the capital of France?"):
        print(response)

if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
```

Output:
```bash
(.venv) ➜  openlit git:(main) ✗ python langchain_with_openai.py
Traceback (most recent call last):
  File "/Users/*********/programming/openlit/langchain_with_openai.py", line 28, in <module>
    asyncio.run(main())
  File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/*********/programming/openlit/langchain_with_openai.py", line 23, in main
    async for response in get_chat_response("What is the capital of France?"):
  File "/Users/*********/programming/openlit/langchain_with_openai.py", line 18, in get_chat_response
    async for response in completion:
  File "/Users/*********/programming/openlit/.venv/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 494, in astream
    raise e
  File "/Users/*********/programming/openlit/.venv/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 472, in astream
    async for chunk in self._astream(
  File "/Users/*********/programming/openlit/.venv/lib/python3.11/site-packages/langchain_openai/chat_models/base.py", line 2038, in _astream
    async for chunk in super()._astream(*args, **kwargs):
  File "/Users/*********/programming/openlit/.venv/lib/python3.11/site-packages/langchain_openai/chat_models/base.py", line 810, in _astream
    async with response:
TypeError: 'async_generator' object does not support the asynchronous context manager protocol
(.venv) ➜  openlit git:(main) ✗
```

Although using openai directly with the `AsyncOpenAI` client works perfectly fine.
```python
import openlit
openlit.init(otlp_endpoint="http://127.0.0.1:4318")


from openai import AsyncOpenAI


client = AsyncOpenAI()

async def get_chat_response(question):
    completion = await client.chat.completions.create(
        messages=[
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": question},
        ],
        model="gpt-3.5-turbo",
        stream=True,
    )
    async for response in completion:
        yield response.choices[0].delta.content


async def main():
    async for response in get_chat_response("What is the capital of France?"):
        print(response)

if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
```


This issue was acknowledged by langchain in this [Github Issue](https://github.com/langchain-ai/langchain/issues/20923), but then dismissed because it wasn't a problem caused by langchain itself, rather than (in the case of the referenced github issue) by the datadog tracing module in this [Github Issue](https://github.com/DataDog/dd-trace-py/issues/9077)

Please let me know, if I should open a separate issue for this matter (I wanted to avoid duplication where possible 😅)

**Comment by amanagarwal042** [comment]
@MohanadAbugharbia 
We can use this issue, Thanks for letting us know. 

Yeah this one surely needs an update, Ill check out what can be done. Gave a few tries on a fix but now sure where the context is getting broken. (any ideas or fixes if you know lemme know or feel free to open a PR)

**Comment by MohanadAbugharbia** [comment]
I believe that the `wrapper` function inside `async_chat_completions` or the `stream_generator` inside the `wrapper` function is the one that needs to support asynchronous context manager protocol, but I have not been able to solve this on my own so far.

It's difficult to solve for me, because I'm trying to wrap my head around all the wrapping that is happening.

**Comment by MohanadAbugharbia** [comment]
Good news, I sat down again and tried different things

Got a working state, now I just have to reintegrate the actual logic or the metrics and traces

Here is the state so far:
inside the `wrapper` function I do the following and return an instance of this `CustomWrapper` class that I created:
```python
if streaming:
    awaited_wrapped = await wrapped(*args, **kwargs)
    return CustomWrapper(awaited_wrapped, instance, *args, **kwargs)
```

Instead of `return stream_generator()`

This is the basic implementation of that class
```python 
class CustomWrapper:
    def __init__(self, wrapped, instance, *args, **kwargs):
        print("CustomWrapper")
        # super().__init__(wrapped)
        self.__wrapped__ = wrapped

    async def __aenter__(self):
        await self.__wrapped__.__aenter__()
        return self

    async def __aexit__(self, exc_type, exc_value, traceback):
        await self.__wrapped__.__aexit__(exc_type, exc_value, traceback)

    def __aiter__(self):
        return self

    async def __anext__(self):
        try:
            chunk = await self.__wrapped__.__anext__()
            return chunk
        except StopAsyncIteration:
            raise StopAsyncIteration
```

It just wraps around the output of the `completions.create` method which is of type `openai.AsyncStream`

Now I have to migrate the logic from the `stream_generator` function into this class and I should be seeing some traces in my local dashboard.

I will update you when I get this done, in the meantime maybe you could let me know if I should do the same solution for other wrappers as well, or if this is enough for now?

P.S. Please excuse the naming of everything in these snippets, I was frustrated with it 😅

**Comment by amanagarwal042** [comment]
I just tested the PR and it looks great!! 

> I will update you when I get this done, in the meantime maybe you could let me know if I should do the same solution for other wrappers as well, or if this is enough for now?

Yeah we should do it for other wrappers/instrumentations as well (Not yet sure if they are affected by the same issue though aswell), Lets do it in a separate PR so this can be released quickly and unblock people affected by this issue

**Comment by amanagarwal042** [comment]
Thanks, everyone for finding the issue and big thanks to @MohanadAbugharbia for implementing it 🙌 

The PR is merged and the fix is available In OpenLIT python SDK version `1.29.2`
